include/libdnf5/base/goal_elements.hpp

```diff
@@ -166,9 +166,12 @@ struct GoalJobSettings : public ResolveSpecSettings {
         return group_package_types ? &group_package_types.value() : nullptr;
     }
 
-    /// If set to true, group operations (install / remove) will only work
+    /// If set to true, group operations (install / remove / upgrade) will only work
     /// with the group itself, but will not add to the transaction any packages.
     bool group_no_packages{false};
+    /// If set to true, environment operations (install / remove / upgrade) will only work
+    /// with the environment itself, but will not add to the transaction any groups.
+    bool environment_no_groups{false};
```

Comment:
This change will brake ABI compatibility (C++).

Reply:
I agree that we should add a pimpl member to every public class on the first need of breaking ABI. So that we can survive without breaking ABI later.

Reply:
I am also in favor of pimpl.

I have a question if we should do it one by one, only when its needed or if we should add the pimpls to all API classes at once preemptively to break the ABI the least amount of times?

Reply:
> I am also in favor of pimpl.
> 
> I have a question if we should do it one by one, only when its needed or if we should add the pimpls to all API classes at once preemptively to break the ABI the least amount of times?

I like the preemption idea. If we can detect all potential places now, I am for preparing the codebase for this purpose in advance.

Reply:
(I have removed the problematic part from the PR so it doesn't block it, there is only a TODO for now.)

Reply:
If we will create `pimpl`, how we can keep API compatibility? The structure does not have methods -> getters, setters